File: library/src/main/java/com/github/dmitrikudrenko/countryphonecodechooser/ui/adapter/CountryCodeSectionAdapter.java
package com.github.dmitrikudrenko.countryphonecodechooser.ui.adapter;

import android.support.annotation.RestrictTo;
import android.support.annotation.RestrictTo.Scope;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.RecyclerView.AdapterDataObserver;
import com.github.dmitrikudrenko.countryphonecodechooser.model.CountryCode;
import com.github.dmitrikudrenko.countryphonecodechooser.ui.adapter.sections.SectionCountry;
import com.github.dmitrikudrenko.countryphonecodechooser.ui.adapter.sections.SectionDataWrapper;

import java.util.List;

@RestrictTo(Scope.LIBRARY)
public abstract class CountryCodeSectionAdapter<T extends RecyclerView.ViewHolder>
        extends RecyclerView.Adapter<T> implements HeaderAdapter {
    protected final CountryCodeAdapter adapter;
    private final SectionDataWrapper sectionDataWrapper = new SectionDataWrapper();

    public CountryCodeSectionAdapter(CountryCodeAdapter adapter) {
        this.adapter = adapter;
        adapter.registerAdapterDataObserver(new AdapterDataObserver() {
            @Override
            public void onChanged() {
                List<CountryCode> codes = adapter.getData();
                sectionDataWrapper.onChanged(codes);
            }
        });
    }

    @Override
    public void setOnCodeSelectListener(final OnCodeSelectListener onCodeSelectListener) {
        adapter.setOnCodeSelectListener(onCodeSelectListener);
    }

    @Override
    public void setSelected(final CountryCode selected) {
        adapter.setSelected(selected);
    }

    @Override
    public void setFilter(final String filter) {
        sectionDataWrapper.setFilter(filter);
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return sectionDataWrapper.getItemCount();
    }

    public SectionCountry getSectionCountry(int position) {
        return sectionDataWrapper.getSectionCountry(position);
    }
}

File: library/src/main/kotlin/com/github/dmitrikudrenko/countryphonecodechooser/ui/adapter/CountryCodeSectionAdapter.kt
package com.github.dmitrikudrenko.countryphonecodechooser.ui.adapter

import android.support.annotation.RestrictTo
import android.support.annotation.RestrictTo.Scope
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.AdapterDataObserver
import com.github.dmitrikudrenko.countryphonecodechooser.model.CountryCode
import com.github.dmitrikudrenko.countryphonecodechooser.ui.adapter.sections.SectionCountry
import com.github.dmitrikudrenko.countryphonecodechooser.ui.adapter.sections.SectionDataWrapper

@RestrictTo(Scope.LIBRARY)
abstract class CountryCodeSectionAdapter<T : RecyclerView.ViewHolder>
internal constructor(protected val adapter: CountryCodeAdapter)
    : RecyclerView.Adapter<T>(), HeaderAdapter {
    private val sectionDataWrapper = SectionDataWrapper()

    init {
        adapter.registerAdapterDataObserver(object : AdapterDataObserver() {
            override fun onChanged() {
                val codes = adapter.data
                sectionDataWrapper.onChanged(codes)
            }
        })
    }

    override fun setOnCodeSelectListener(onCodeSelectListener: OnCodeSelectListener) {
        adapter.setOnCodeSelectListener(onCodeSelectListener)
    }

    override fun setSelected(selected: CountryCode) {
        adapter.setSelected(selected)
    }

    override fun setFilter(filter: String) {
        sectionDataWrapper.setFilter(filter)
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return sectionDataWrapper.itemCount
    }

    internal fun getSectionCountry(position: Int): SectionCountry {
        return sectionDataWrapper.getSectionCountry(position)
    }
}
